### PR TITLE
Add std.meta.applyLeft and std.meta.applyRight

### PR DESCRIPTION
**Pull request by JakobOvrum**
This should help eliminate a lot of tiny helper templates, including those that tend to be visible in template constraints despite not being documented anywhere.

Ping @Dicebot, @John-Colvin in case there's something about std.meta I should know about before adding new functionality to it.

### CODE REVIEW

File: std/meta.d [hunk 1]
```diff
@@ -954,6 +954,98 @@ unittest
...
+        }
+
+        Test foo;
+    }
+
+    static assert(allSatisfy!(applyRight!(hasMember, "foo"), T1, T2));
+    static assert(allSatisfy!(applyRight!(ifTestable, a => a.foo), T1, T2));
+}
+
+// Tests for enum/value results
```
Comment (JakobOvrum) [inline std/meta.d]: Oops, what I wanted to test here was type/symbol results. Will rewrite this test.

**Comment by JakobOvrum** [comment]
The unit tests now also test alias results.

**Comment by JackStouffer** [comment]
LGTM

File: std/meta.d [hunk 2]
```diff
@@ -954,6 +954,117 @@ unittest
...
+  * Behaves like the identity function when $(D_PARAM args) is empty.
+  * Params:
+  *    Template = template to partially apply
+  *    args     = arguments to bind
+  * Returns:
+  *    _Template with arity smaller than or equal to $(D_PARAM Template)
+  */
+template applyLeft(alias Template, args...)
+{
+    static if(args.length)
```
Comment (quickfur) [inline std/meta.d]: Nitpick: space between `if` and `(`.

File: std/meta.d [hunk 3]
```diff
@@ -954,6 +954,117 @@ unittest
...
+        }
+    }
+    else
+        alias applyLeft = Template;
+}
+
+/// Ditto
+template applyRight(alias Template, args...)
+{
+    static if(args.length)
```
Comment (quickfur) [inline std/meta.d]: ditto

File: std/meta.d [hunk 4]
```diff
@@ -954,6 +954,117 @@ unittest
...
+  * Returns:
+  *    _Template with arity smaller than or equal to $(D_PARAM Template)
+  */
+template applyLeft(alias Template, args...)
+{
+    static if(args.length)
+    {
+        template applyLeft(right...)
+        {
+            static if(is(typeof(Template!(args, right))))
```
Comment (quickfur) [inline std/meta.d]: ditto

File: std/meta.d [hunk 5]
```diff
@@ -954,6 +954,117 @@ unittest
...
+}
+
+/// Ditto
+template applyRight(alias Template, args...)
+{
+    static if(args.length)
+    {
+        template applyRight(left...)
+        {
+            static if(is(typeof(Template!(left, args))))
```
Comment (quickfur) [inline std/meta.d]: ditto

**Comment by quickfur** [comment]
Other than that, LGTM.

**Comment by John-Colvin** [comment]
LGTM, not sure why I never thought to write these myself.

File: std/meta.d (hunk 5 above)
Comment (John-Colvin) [inline std/meta.d]: space after `if`

File: std/meta.d (hunk 4 above)
Comment (John-Colvin) [inline std/meta.d]: space after `if`

**Comment by John-Colvin** [comment]
@JakobOvrum any chance you could fix those missing spaces? I'd love to see this merged.

**Comment by JakobOvrum** [comment]
Fixed the spaces, but of course anyone with merge rights could just merge and fix such tiny issues themselves to speed up the process.

**Comment by quickfur** [comment]
Auto-merge toggled on

**Comment by quickfur** [comment]
Let's just get the queue moving again, and leave minor improvements to followup PRs.

**Pull request merged**